controllers/metal3.io/hostfirmwaresettings_controller.go

```diff
@@ -0,0 +1,466 @@
+		firmwareSchema.Spec.HardwareModel = info.bmh.Status.HardwareDetails.SystemVendor.ProductName
+	}
+
+	// Copy in the schema from provisioner
+	firmwareSchema.Spec.Schema = make(map[string]metal3v1alpha1.SettingSchema)
+	for k, v := range schema {
+		firmwareSchema.Spec.Schema[k] = v
+	}
+
+	if err = r.Create(context.TODO(), firmwareSchema); err != nil {
+		return nil, err
+	}
+	// Set hfs as owner after the create
+	if controllerutil.SetOwnerReference(info.hfs, firmwareSchema, r.Scheme()); err != nil {
+		return nil, errors.Wrap(err, "could not set owner of firmwareSchema")
+	}
+	if err = r.Update(context.TODO(), firmwareSchema); err != nil {
+		return nil, err
+	}
+
+	info.log.Info("created new firmwareSchema resource")
+
+	return firmwareSchema, nil
+}
+
+// Create a hostFirmwareSettings
+func (r *HostFirmwareSettingsReconciler) newHostFirmwareSettings(info *rInfo, namespacedName types.NamespacedName) (err error) {
+
+	info.hfs.ObjectMeta = metav1.ObjectMeta{
+		Name: namespacedName.Name, Namespace: namespacedName.Namespace}
+	info.hfs.Status.Settings = make(metal3v1alpha1.SettingsMap)
+	info.hfs.Spec.Settings = make(metal3v1alpha1.DesiredSettingsMap)
+
+	if err = r.Create(context.TODO(), info.hfs); err != nil {
+		return errors.Wrap(err, "failure creating hostFirmwareSettings resource")
+	}
+
+	// Set bmh as owner, this makes sure the resource is deleted when bmh is deleted
+	if controllerutil.SetControllerReference(info.bmh, info.hfs, r.Scheme()); err != nil {
+		return errors.Wrap(err, "could not set bmh as controller")
+	}
+	if err = r.Update(context.TODO(), info.hfs); err != nil {
+		return errors.Wrap(err, "could not update hostfirmwaresettings")
+	}
+
+	info.log.Info("created new hostFirmwareSettings resource")
+
+	return nil
+}
+
+// EventHandler for updates to both the baremetalhost and hostfirmwarettings
+func (r *HostFirmwareSettingsReconciler) updateEventHandler(e event.UpdateEvent) bool {
+
+	r.Log.Info("hostfirmwaresettings in event handler")
+
+	// If this is a baremetalhost, only return true on certain state transitions
+	var oldState metal3v1alpha1.ProvisioningState = metal3v1alpha1.StateNone
+	newState := oldState
+
+	oldHost, oldHostExists := e.ObjectOld.(*metal3v1alpha1.BareMetalHost)
+	if oldHostExists {
+		oldState = oldHost.Status.Provisioning.State
+	}
+	newHost, newHostExists := e.ObjectNew.(*metal3v1alpha1.BareMetalHost)
+	if newHostExists {
+		newState = newHost.Status.Provisioning.State
+	}
+
+	if newHostExists && !oldHostExists {
+		// The baremetalhost has been created
+		return true
+	}
+	if oldHostExists || newHostExists {
+		// Data needs to be retrieved from Ironic when node is moved to Ready  e.g. after cleaning and when
+		// provisioned, its not necessary on every state change as data from Ironic will be the same.
+		if (newState == metal3v1alpha1.StateReady || newState == metal3v1alpha1.StateProvisioned) && oldState != newState {
```

Comment:
I don't believe it's safe in general to try to build an edge-triggered controller like this. Kubernetes is designed for everything to be [level-triggered](https://hackernoon.com/level-triggering-and-reconciliation-in-kubernetes-1f17fe30333d). There is no guarantee that we will see the edge (e.g. if the controller is restarted).

If we have to trigger on an edge I think the only safe way of doing so is to run it inside the BaremetalHost reconciler before writing the new provisioning status.

Reply:
`StateReady` is being renamed to `StateAvailable` (#340), so wherever one is referenced we have to include the other.

Reply:
Will change to remove the edge-triggering in a follow on PR.